processing/src/main/java/org/apache/druid/segment/data/FrontCodedIndexed.java

```diff
@@ -0,0 +1,393 @@
+          copy.position(bucketsPosition + offset);
+          currentBucket = readBucket(
+              copy,
+              bucketNum < (numBuckets - 1) ? bucketSize : lastBucketNumValues
+          );
+          currentBucketIndex = bucketNum;
+        }
+        int offset = adjustedCurrIndex & rem;
+        currIndex++;
+        return currentBucket[offset];
+      }
+
+      @Override
+      public void remove()
+      {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  @Override
+  public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  {
+    inspector.visit("buffer", buffer);
+    inspector.visit("hasNulls", hasNull);
+    inspector.visit("bucketSize", bucketSize);
+  }
+
+  /**
+   * Get a value from a bucket at a relative position
+   */
+  public static ByteBuffer getFromBucket(ByteBuffer bucket, int offset)
+  {
+    final byte[] first = readBytes(bucket);
+    if (offset == 0) {
+      return ByteBuffer.wrap(first);
+    }
+    int pos = 0;
+    int prefixLength;
+    final byte[] fragment;
+    // scan through bucket values until we reach offset
+    do {
+      prefixLength = VByte.readInt(bucket);
+      if (++pos < offset) {
+        final int skipLength = VByte.readInt(bucket);
+        bucket.position(bucket.position() + skipLength);
+      } else {
+        fragment = readBytes(bucket);
+        break;
+      }
+    } while (true);
+    return combinePrefixAndFragment(first, prefixLength, fragment);
+  }
+
+  /**
+   * Find the relative position of a value in a bucket with a linear scan
+   */
+  public static int findInBucket(ByteBuffer bucket, int numValues, ByteBuffer value, Comparator<ByteBuffer> comparator)
+  {
+    final byte[] first = readBytes(bucket);
+    final int firstCompare = comparator.compare(ByteBuffer.wrap(first), value);
+    if (firstCompare == 0) {
+      return 0;
+    }
+    if (firstCompare > 0) {
+      return -1;
+    }
+    int offset = 0;
+    int prefixLength;
+    byte[] fragment;
+    // scan through bucket values until we find match or compare numValues
+    int insertionPoint = 1;
+    while (++offset < numValues) {
+      prefixLength = VByte.readInt(bucket);
+      fragment = readBytes(bucket);
+      ByteBuffer next = combinePrefixAndFragment(first, prefixLength, fragment);
+      final int cmp = comparator.compare(next, value);
+      if (cmp == 0) {
+        return offset;
+      } else if (cmp < 0) {
+        insertionPoint++;
+      } else {
+        break;
+      }
+    }
+    return -(insertionPoint + 1);
+  }
+
+  /**
+   * reconstruct a value given the bytes of the first bucket value, the length of the first value to use as a prefix,
+   * and the fragment of the value to use after the prefix.
+   */
+  private static ByteBuffer combinePrefixAndFragment(byte[] first, int prefixLength, byte[] fragment)
+  {
+    ByteBuffer next = ByteBuffer.allocate(prefixLength + fragment.length);
```

Comment:
In reading this code, I find myself wondering if this class wouldn't be better off with a Buffer that it reuses and callers of this class needs to be aware that the Buffer gets reused (so they must make a copy of the contents if they want to hang onto a reference of it).  

In general, the buffers returned from things get reused, so I'm pretty sure most of the client code should be fine with this semantic (e.g. the `String` indexed wrapper would consume and more on really fast) and it would reduce object allocations.

Reply:
I thought about this, but was unsure if this would be ok and didn't want to put in the work to find out as part of this PR; `GenericIndexed` with the buffer strategy uses `asReadOnlyBuffer` on get, which also allocates new buffers, so it isn't something that is known to be cool.

I guess to share a buffer, we would need to allocate one with the maximum string length, which i should probably store in the dictionary itself if we are sure we want to do this, else it needs to be growable